execution/sym_region_frag_machine.ml

```diff
@@ -1189,8 +1191,37 @@ struct
 	  Printf.printf "Largest value based on queries is %Ld\n" limit;
 	limit
 
+    method private query_minval e ty =
+      let rec loop min max =
+	assert(min <= max);
+	if min = max then
+	  min
+	else
+	  let mid =
+	    Int64.shift_right (Int64.succ (Int64.add min max)) 1
+	  in
+	  let cond_e = V.BinOp(V.SLE, V.Constant(V.Int(ty, mid)), e) in
+	  let in_bounds = self#query_valid cond_e in
+	    if !opt_trace_tables then
+	      Printf.printf "(%s) >=$ %Ld: %s\n" (V.exp_to_string e) mid
+		(if in_bounds then "valid" else "invalid");
+	    if in_bounds then
+	      loop mid max 
+	    else
+	      loop min (Int64.pred mid)
+      in
+      let wd = min 62 (narrow_bitwidth_signed form_man e) in
+      let max_limit = Int64.shift_right_logical (-1L) (64-wd) in
+      let min_limit = Int64.pred (Int64.neg max_limit) in
+      let limit = loop min_limit max_limit in
+	if !opt_trace_tables then
+	  Printf.printf "Smallest value based on queries is %Ld\n" limit;
+	limit
+
     val maxval_cache = Hashtbl.create 101
     val maxval_offset_cache = Hashtbl.create 101
+    val minval_cache = Hashtbl.create 101
```

Comment:
Here you're creating the data structure for a cache, but it looks like you're not using it.

I think caching minvals would help performance, probably sometimes even more than the more interesting optimizations you've been looking at. But the cache data structures and the code to use them should be together in one commit, so either implement a cache now, or remove the new data structures for now and add them once there's also code to use them.